Return more useful error information when a persistent volume fails to mount

screeley44 opened the pull request:
Fixes #22992  
Fixes #23048   - resolves following issues - UXP issues

volume plugin type errors are not being surfaced to pod describe events properly when a failure occurs with PV/PVC volume building - for example plugin builder fails if secrets or endpoints are missing (rbd or glusterfs).  The user only sees a generic error that has incorrect meaning "unsupported volume type"

see issues  #22992 and #23048 for more details

screeley44 committed:
reword volume mount errUnsupportedVolume error and surface true error to describe event

screeley44 commented:
@smarterclayton @pmorie @pweil-  please review

pweil- commented on pkg/kubelet/volumes.go:
```diff
@@ -34,7 +35,7 @@ import (
 )
 
-var errUnsupportedVolumeType = fmt.Errorf("unsupported volume type")
+var errUnsupportedVolumeType = fmt.Errorf("failed volume mount, check previous pod describe events or logs for more details")
```
@screeley44 @pmorie - I took a look at the usages of this error.  It looks like some of the usages (`NewWrapperBuilder`, `NewWrapperCleaner`) will actually hide the underlying error from the `New*` call and some don't (`mountExternalVolumes`).  Should we go ahead and standardize that too?  

I'll defer to folks who are in this code more than me.

pmorie commented:
There are a few different problems to untangle here:

1.  `NewWrapperBuilder` and `NewWrapperCleaner` in `volumes.go` both swallow errors when making their respective result objects; we should instead be passing those back
2.  `newVolumeBuilderFromPlugins` should return a nil and an error, or a builder and no error, and not locating a volume plugin should be considered an error.  Remove the check for a nil builder in `mountExternalVolumes`
3.  `unsupported volume type` doesn't mean anything to anyone -- the error message for this case should be something like "Unable to find a volume plugin that can handle volume of type X` -- there's some overlap with the stuff @pweil is doing with volumes and OpenShift SCCs (which is not yet ported upstream) that has some code for getting user-facing labels from Volume specs that could be used to provide a meaningful guess at the volume type, but I think in practice this error message should almost never happen.

I believe if we make these changes both the error message in the pod status and the message in the event will have the accurate root cause.

How does that sound @screeley44 @pweil-?

cc @saad-ali @thockin

pweil- commented:
> NewWrapperBuilder and NewWrapperCleaner in volumes.go both swallow errors when making their respective result objects; we should instead be passing those back

Yes, agree

> newVolumeBuilderFromPlugins should return a nil and an error, or a builder and no error, and not locating a volume plugin should be considered an error. Remove the check for a nil builder in mountExternalVolumes

I like this better.  I noticed in the review (https://github.com/kubernetes/kubernetes/pull/23122#discussion_r56539362) that we have to go through extra gyrations because methods can return `nil, nil` which is "acceptable" but then just turned into an error anyway.  I'd rather bubble a "plugin not registered" or similar error and change the calling code to be more simple.

saad-ali commented:
:+1: to @pmorie's suggestions

dchen1107 commented:
+1 on @pmorie's comment: https://github.com/kubernetes/kubernetes/pull/23122#issuecomment-198203694

@screeley44 Are you going to continue with pr with @pmorie's suggestion?

screeley44 commented:
@dchen1107 @pmorie -  Yes, I would like to continue, I've been off for spring break and Easter but back now and plan on picking back up on this

screeley44 commented:
@pmorie  @pweil-    please review what I did, I think this will expose all root cause errors to the user in the describe event as I tried to thoroughly test all conditions, but maybe I missed something or could have done something in a better way.

```
Events:
  FirstSeen	LastSeen	Count	From			SubobjectPath	Type		Reason		Message
  ---------	--------	-----	----			-------------	--------	------		-------
  12s		12s		1	{default-scheduler }			Normal		Scheduled	Successfully assigned bb-gluster-pod2 to 127.0.0.1
  12s		1s		2	{kubelet 127.0.0.1}			Warning		FailedMount	Unable to mount volumes for pod "bb-gluster-pod2_default(9fd3f3a2-f61a-11e5-b038-52540092b5fb)": failed to instantiate volume: glustervol for plugin: kubernetes.io/persistent-claim with root cause: failed to instantiate volume: pv-gluster for plugin: kubernetes.io/glusterfs with root cause: endpoints "glusterfs-cluster" not found
```

I also removed the errUnsupportedVolumeType generic error and added a method, I tried to replicate error scenarios where that might get called, but I really couldn't find any, but maybe doesn't hurt to keep around just in case?

screeley44 commented:
@dchen1107  -  any thoughts on this so far?

erinboyd commented:
@pmorie @saad-ali @pweil- @smarterclayton where are we with this? Can we please merge this before ANOTHER rebase?

smarterclayton commented:
@kubernetes/sig-storage

swagiaal commented on pkg/kubelet/volumes.go:
```diff
@@ -34,8 +34,6 @@ import (
 )
 
-var errUnsupportedVolumeType = fmt.Errorf("unsupported volume type")
-
```
Hmm.. do we really not need this or can we move it somewhere where we are sure that the volume is not supported ?

screeley44 replied:
this is removed and replaced with a more generic error  createVolumeTypeError - which in theory should never be called from what I can tell, so I was asking if we can remove it all together (errUnsupportedVolumeType was original and I removed in this PR) but replaced with a function to eventually help generate a better error...but as I wondered in the PR I don't know if that error would ever get hit, since if the mounter fails in anyway that error is now surfaced where as before it wasn't...so wonder if we can remove altogether or keep around for some unforeseen scenario??  @swagiaal

swagiaal replied:
Okay I am convinced it is not needed after reading the rest of the patch :)

pmorie replied:
It should not be needed, and there's no utility to keeping it around 'just in case'.  YAGNI

swagiaal commented on pkg/kubelet/volumes.go:
```diff
@@ -66,10 +64,16 @@ func (vh *volumeHost) NewWrapperMounter(volName string, spec volume.Spec, pod *a
+	}
+
+	// generic volume type error
+	return nil, createVolumeTypeError(volName)
```
can the above be replaced with:
```go
return vh.kubelet.newVolumeMounterFromPlugins(&spec, pod, opts)
```
?

swagiaal commented on pkg/kubelet/volumes.go:
```diff
@@ -87,11 +91,22 @@ func (vh *volumeHost) NewWrapperUnmounter(volName string, spec volume.Spec, podU
-	return c, nil
+
+	// generic volume type error
+	return nil, createVolumeTypeError(volName)
```
Same comment as above

pmorie replied:
I have same concern as I articulated above.  `NewUnmounter` should never, ever return (nil, nil).

swagiaal commented on pkg/kubelet/volumes.go:
```diff
@@ -87,11 +91,22 @@ func (vh *volumeHost) NewWrapperUnmounter(volName string, spec volume.Spec, podU
+}
+
+func createVolumeTypeError(volName string) error {
+	return fmt.Errorf("Unable to mount volume or unsupported volume type for volume: %s for plugin: Unknown - check logs or pod describe events for possible root cause", volName)
```
Hmm... I am not sure if we should prefix errors with all of that.
For example I think that:

`endpoints "glusterfs-cluster" not found`

provides better use experience than:

`FailedMount Unable to mount volumes for pod "bb-gluster-pod2_default(9fd3f3a2-f61a-11e5-b038-52540092b5fb)": failed to instantiate volume: glustervol for plugin: kubernetes.io/persistent-claim with root cause: failed to instantiate volume: pv-gluster for plugin: kubernetes.io/glusterfs with root cause: endpoints "glusterfs-cluster" not found`

Then again the underlying error might require explanation. Here I think we can get away with not prefixing and just returning the error. If the underlying error is too cryptic we can fix it at the point where it is generated.

screeley44 replied:
yeah, so what you are seeing here, is the result of two volume types producing errors, if the user was not using PVC/PV then you would only see the error associated with the underlying source volume (i.e. glusterfs type) but with PVC and PV types you also get another type of persistent-claim with no real error other than it fails but the real error comes from the underlying source plugin type.  So I guess we could repress the generic persistent-claim error and only show the underlying issue from the source type??

swagiaal replied:
> yeah, so what you are seeing here, is the result of two volume types producing errors, if the user was not using PVC/PV then you would only see the error associated with the underlying source volume (i.e. glusterfs type with the true root cause of missing endpoints) ... but with PVC and PV types you also get another type of persistent-claim with no real error other than it fails at the mounter level but the real error comes from the underlying source plugin type.

Gotcha!

> So I guess we could repress the generic persistent-claim error and only show the underlying issue from the source type??

I think that would be right thing to do. Not prefix errors unless they are too cryptic so that the use would not have to parse through a lot of boiler plate to get to the real issue. I would do that in a separate PR though or open an issue so it is not forgotten.

The instance in this PR though should be addressed here.

pmorie replied:
We need to expose the actual error wherever possible; don't eat errors, and don't add vague text to errors.  Think about it from the user's perspective -- we want to give them the information most relevant to solving the problem.

pmorie replied:
To make it explicit, we should not need this error.

swagiaal commented on pkg/kubelet/volumes.go:
```diff
@@ -132,8 +147,9 @@ func (kl *Kubelet) mountExternalVolumes(pod *api.Pod) (kubecontainer.VolumeMap,
 			glog.Errorf("Could not create volume mounter for pod %s: %v", pod.UID, err)
 			return nil, err
 		}
+		//This should never be hit if there is a failure in mounter - probably can remove??
```
I think it is reachable unless I am missing something.
`newVolumeMounterFromPlugins` can return `nil, nil` although `FindPluginBySpec` which it calls never returns `nil, nil`
I think it is safe to remove but might as well leave it for completeness. Eitherway I would remove the comment above

swagiaal commented on pkg/kubelet/volumes.go:
```diff
@@ -132,8 +147,9 @@ func (kl *Kubelet) mountExternalVolumes(pod *api.Pod) (kubecontainer.VolumeMap,
+		//This should never be hit if there is a failure in mounter - probably can remove??
 		if mounter == nil {
-			return nil, errUnsupportedVolumeType
+			return nil, createVolumeTypeError(internal.Volume.Name)
```
I guess this would be an unknown error because when the plugin is not found you usually get an error.

swagiaal commented on pkg/kubelet/volumes.go:
```diff
@@ -243,7 +259,7 @@ func (kl *Kubelet) getPodVolumesFromDisk() map[string]cleanerTuple {
 			}
 			if unmounter == nil {
-				glog.Errorf("Could not create volume unmounter for %s: %v", volume.Name, errUnsupportedVolumeType)
+				glog.Errorf("Could not create volume unmounter for %s: %v", volume.Name, createVolumeTypeError(volume.Name))
```
This is also an unknown error because when the plugin is not found you get an error message.

screeley44 replied:
yeah, and now that wrappers are not eating errors, I think this would most likely never get called since the err would not be nil and it would be surfaced, I left it just because of cases that *might* happen. 
The idea behind the "createVolumeTypeError" was to replace the errUnsupportedVolumeType - which really meant nothing and was inaccurate, so I wanted a "generic" error for these unknown conditions that might pop up and wanted to find a way to give the user some more useful information (i.e. name of the volume from the spec and ideally the acutal plugin type - but plugin type was proving difficult, so figured at least giving the volume.Name was more useful for users than an error saying "Volume Type is Not Supported"   @swagiaal

swagiaal replied:
That makes sense.. especially after rereading the error message :). I am ok with this as is

pmorie replied:
We should not need to check for a nil unmounter.  Never return (nil, nil) from a method like this.

swagiaal commented on pkg/kubelet/volumes.go:
```diff
@@ -274,7 +290,7 @@ func (kl *Kubelet) newVolumeMounterFromPlugins(spec *volume.Spec, pod *api.Pod,
 	physicalMounter, err := plugin.NewMounter(spec, pod, opts)
 	if err != nil {
-		return nil, fmt.Errorf("failed to instantiate volume physicalMounter for %s: %v", spec.Name(), err)
+		return nil, fmt.Errorf("failed to instantiate volume: %s for plugin: %s with root cause: %v", spec.Name(), plugin.Name(), err)
```
physicalMounter is accurate here.. please leave it in

pmorie commented on pkg/kubelet/volumes.go:
```diff
@@ -66,10 +64,16 @@ func (vh *volumeHost) NewWrapperMounter(volName string, spec volume.Spec, pod *a
 	}
 
 	b, err := vh.kubelet.newVolumeMounterFromPlugins(&spec, pod, opts)
```
@screeley44 see my earlier feedback: `newVolumeMounterFromPlugins` should return a mounter and a nil error, or a nil mounter and a non-nil error, but never (nil, nil)

pmorie replied:
`newVolumeMounterFromPlugins` should return an error if there's no suitable volume plugin.

screeley44 replied:
no more nil, nil being returned from any of the functions - also no errors are being eaten, so any issue will be surfaced - think this addresses your concern @pmorie

pmorie replied:
let's rename `b` to `mounter`

pmorie commented on pkg/kubelet/volumes.go:
```diff
@@ -268,13 +272,9 @@ func (kl *Kubelet) newVolumeMounterFromPlugins(spec *volume.Spec, pod *api.Pod,
 	physicalMounter, err := plugin.NewMounter(spec, pod, opts)
 	if err != nil {
-		return nil, fmt.Errorf("failed to instantiate volume physicalMounter for %s: %v", spec.Name(), err)
+		return nil, fmt.Errorf("failed to instantiate volume physicalMounter for: %s for plugin: %s with root cause: %v", spec.Name(), plugin.Name(), err)
```
`physicalMounter for volume:` plz

pmorie commented on pkg/kubelet/volumes.go:
```diff
@@ -285,10 +285,6 @@ func (kl *Kubelet) newVolumeAttacherFromPlugins(spec *volume.Spec, pod *api.Pod,
 	if err != nil {
 		return nil, fmt.Errorf("can't use volume plugins for %s: %v", spec.Name(), err)
 	}
-	if plugin == nil {
```
I think this has to stay because not all plugins are necessarily attachable.

markturansky replied:
like @pmorie's comment on Detacher below, this is also valid.  Not every volume plugin will have an attacher.

markturansky replied:
Correct for the Attachable bit.  This is for the volume plugin itself.

I think returning `nil, nil` is an error at this point.  No plugin was found to satisfy the volume spec.

swagiaal replied:
I think it is okay to remove this because FindPluginBySpec does not return nil,nil. If you did not get an error you are getting a plugin.

markturansky replied:
I don't know how I cross reference morie's comment.  he's got it right.  

@swagiaal i agree with you.

pmorie replied:
Yes, sami is right, and this is fine the way it is now.

pmorie commented on pkg/kubelet/volumes.go:
```diff
@@ -83,15 +87,22 @@ func (vh *volumeHost) NewWrapperUnmounter(volName string, spec volume.Spec, podU
-		return nil, nil
-	}
+
 	c, err := plugin.NewUnmounter(spec.Name(), podUID)
```
I think it would be an improvement if `c` was renamed to `unmounter`

pmorie commented on pkg/kubelet/volumes.go:
```diff
@@ -66,10 +64,16 @@ func (vh *volumeHost) NewWrapperMounter(volName string, spec volume.Spec, pod *a
+		return nil, err
 	}
-	return b, nil
+	if b != nil {
```
This check _should_ be unnecessary now.

pmorie commented on pkg/kubelet/volumes.go:
```diff
@@ -323,10 +317,6 @@ func (kl *Kubelet) newVolumeDetacherFromPlugins(kind string, name string, podUID
 	if err != nil {
 		return nil, fmt.Errorf("can't use volume plugins for %s/%s: %v", podUID, kind, err)
 	}
-	if plugin == nil {
```
This is also a valid state because not every volume has a detacher

markturansky commented on pkg/kubelet/volumes.go:
```diff
@@ -300,15 +296,13 @@ func (kl *Kubelet) newVolumeAttacherFromPlugins(spec *volume.Spec, pod *api.Pod,
 
 func (kl *Kubelet) newVolumeUnmounterFromPlugins(kind string, name string, podUID types.UID) (volume.Unmounter, error) {
 	plugName := strings.UnescapeQualifiedNameForDisk(kind)
+	glog.Infof("--- Scott - plugName = %s", plugName)
```
debug statement

swagiaal commented on pkg/kubelet/volumes.go:
```diff
@@ -65,11 +63,13 @@ func (vh *volumeHost) NewWrapperMounter(volName string, spec volume.Spec, pod *a
 	}
-	return b, nil
+
+	return mounter, nil
```
you don't need to check the return values here if you are just returning them. Just `return vh.kubelet.newVolumeMounterFromPlugins(&spec, pod, opts)` should do it

pmorie replied:
yeah, that's a good point.

screeley44 replied:
yeah, agree, sorry I missed your original comment @swagiaal  -  these PRs get messy

swagiaal replied:
> sorry I missed your original comment @swagiaal - these PRs get messy

no worries... almost there @screeley44!

swagiaal commented on pkg/kubelet/volumes.go:
```diff
@@ -79,19 +79,20 @@ func (vh *volumeHost) NewWrapperUnmounter(volName string, spec volume.Spec, podU
-	return c, nil
+
+	return unmounter, nil
 }
```
Same here unless I am missing something.
`return plugin.NewUnmounter(spec.Name(), podUID)`

pmorie replied:
You're not missing anything, good point.

swagiaal commented on pkg/kubelet/volumes.go:
```diff
@@ -79,19 +74,15 @@ func (vh *volumeHost) NewWrapperUnmounter(volName string, spec volume.Spec, podU
 	}
 
+	//FindPluginBySpec will only return a valid plugin if only one matched
+	// or an error if 0 plugins matched or more than 1 plugin matched
```
These comments and the others like it below don't belong here. You should add them to the godoc of the function being called (FindPluginBySpec, NewUnmounter, newVolumeMounterFromPlugins, etc)

pmorie replied:
Agree with @swagiaal.  Also, please add Godoc comments for `NewWrapperUnmounter` and friends (not just the ones you touch in this PR).

pmorie commented on pkg/kubelet/volumes.go:
```diff
@@ -264,23 +248,22 @@ func (kl *Kubelet) getPodVolumesFromDisk() map[string]cleanerTuple {
 func (kl *Kubelet) newVolumeAttacherFromPlugins(spec *volume.Spec, pod *api.Pod, opts volume.VolumeOptions) (volume.Attacher, error) {
+	// FindAttachablePluginBySpec can return an Attacher or an error or a nil detacher and error
+	// as not all volumes will have an attacher/detacher
 	plugin, err := kl.volumePluginMgr.FindAttachablePluginBySpec(spec)
```
This needs to move to the definition of `FindAttachablePluginBySpec`

screeley44 replied:
@pmorie - I removed from here as the def of FindAttachablePluginBySpec already has a good godoc

pmorie commented on pkg/kubelet/volumes.go:
```diff
@@ -263,23 +251,27 @@ func (kl *Kubelet) getPodVolumesFromDisk() map[string]cleanerTuple {
 	return currentVolumes
 }
 
+// newVolumeMounterFromPlugins attempts to find a plugin by volume spec and then create a physicalMounter from
```
Strike this comment; there's a lot more to say about this method than is in-scope for this PR

pmorie commented on pkg/kubelet/volumes.go:
```diff
@@ -263,23 +251,27 @@ func (kl *Kubelet) getPodVolumesFromDisk() map[string]cleanerTuple {
 }
 
+// newVolumeAttacherFromPlugins attempts to find a plugin by a persistent volume spec and then create an attacher.
+// This is used when a volume needs to attach to a node before mounting.
```
Strike this line

pmorie commented on pkg/kubelet/volumes.go:
```diff
@@ -263,23 +251,27 @@ func (kl *Kubelet) getPodVolumesFromDisk() map[string]cleanerTuple {
 
+// newVolumeAttacherFromPlugins attempts to find a plugin by a persistent volume spec and then create an attacher.
+// This is used when a volume needs to attach to a node before mounting.
+// This func will return an error if the attacher plugin was not found or if the Attacher failed to instantiate.
```
The information in this line and below is appropriate content, the tense is a little off.  Use present tense, viz: 'returns an error '

pmorie commented on pkg/kubelet/volumes.go:
```diff
@@ -298,6 +290,9 @@ func (kl *Kubelet) newVolumeAttacherFromPlugins(spec *volume.Spec, pod *api.Pod,
 
+// newVolumeUnmounterFromPlugins attempts to find a plugin by name and then create an unmounter from
+// a recoverable state.
+// This will return an error or a valid unmounter only
```
Returns a valid unmounter or an error.

screeley44 commented:
@pmorie @swagiaal - how does this look?

swagiaal commented on pkg/kubelet/volumes.go:
```diff
@@ -106,11 +99,13 @@ func (vh *volumeHost) GetWriter() io.Writer {
 	return vh.kubelet.writer
 }
 
-// Returns the hostname of the host kubelet is running on
```
Why are you removing this ?.. is it not true ?

screeley44 replied:
@pmorie  previous comment said those are out of scope and I originally godoc'd all of them but he said they should be removed for now - although I think that Paul comment was in slack an not an official comment in this PR - that one line you reference above existed before I did all the other Get* functions, so I just removed it as well to be consistent...will put back in if you think I should @swagiaal @pmorie

swagiaal replied:
Hmm.. I think it should only be removed if it is not true anymore. @pmorie can you clarify ?

Also a general rule we try to keep a PR or a commit focused on related changes so if someone uses `git blame` in the future they should be able to understand why a line was changed when looking at the overall commit.

screeley44 replied:
@swagiaal @pmorie  -  gottcha, and that makes sense.  I added back for now, unless Paul has a better reason to remove it.

swagiaal commented on pkg/kubelet/volumes.go:
```diff
@@ -305,10 +295,8 @@ func (kl *Kubelet) newVolumeUnmounterFromPlugins(kind string, name string, podUI
-		return nil, nil
-	}
+
+	// Can only return a valid unmounter or an error
```
This should move to the godoc of NewUnmounter.

swagiaal commented:
@screeley44 it is looking good just a couple of nits.

pmorie commented on pkg/kubelet/volumes.go:
```diff
@@ -65,13 +66,12 @@ func (vh *volumeHost) NewWrapperMounter(volName string, spec volume.Spec, pod *a
 }
 
+// NewWrapperUnmounter attempts to create a volume unmounter from a volume name and pod uid.
+// All volumes have a mounter and unmounter.
```
I think you can strike this line

pmorie commented on pkg/kubelet/volumes.go:
```diff
@@ -58,6 +56,9 @@ func (vh *volumeHost) GetKubeClient() clientset.Interface {
 }
 
+// NewWrapperMounter attempts to create a volume mounter from a volume Spec, pod and volume options.
+// All volumes have a mounter and unmounter.
```
I think you can strike this line

pmorie commented on pkg/kubelet/volumes.go:
```diff
@@ -111,6 +104,9 @@ func (vh *volumeHost) GetHostName() string {
 	return vh.kubelet.hostname
 }
 
+// mountExternalVolumes creates a VolumeMap and attempts to physically mount
```
strike this comment, I made another PR adding a comment for this method (there's a lot more to say about it than this)

pmorie commented on pkg/kubelet/volumes.go:
```diff
@@ -268,18 +257,17 @@ func (kl *Kubelet) newVolumeMounterFromPlugins(spec *volume.Spec, pod *api.Pod,
 	return physicalMounter, nil
 }
 
+// newVolumeAttacherFromPlugins attempts to find a plugin by a persistent volume spec and then create a volume Attacher.
```
s/persistent//

pmorie replied:
Justify comments at 80 characters per line, please

pmorie commented on pkg/kubelet/volumes.go:
```diff
@@ -268,18 +257,17 @@ func (kl *Kubelet) newVolumeMounterFromPlugins(spec *volume.Spec, pod *api.Pod,
 
+// newVolumeAttacherFromPlugins attempts to find a plugin by a persistent volume spec and then create a volume Attacher.
+// This function returns a valid volume Attacher or an error if the Attacher plugin was not found or if the Attacher failed to instantiate.
+// It can also return nil, nil as not all volumes will have an attacher/detacher
```
I'd combine this and the previous sentence:

```
Returns:
- an attacher if one was found
- an error if no plugin could be found for the volume or the attacher failed to instantiate
- nil if there is no appropriate attacher for this volume

pmorie commented on pkg/kubelet/volumes.go:
```diff
@@ -298,6 +286,9 @@ func (kl *Kubelet) newVolumeAttacherFromPlugins(spec *volume.Spec, pod *api.Pod,
 }
 
+// newVolumeUnmounterFromPlugins attempts to find a plugin by name and then create an Unmounter from
+// a recoverable state.
```
I don't understand what the 'recoverable state' is referring to here.

pmorie commented on pkg/kubelet/volumes.go:
```diff
@@ -317,6 +305,9 @@ func (kl *Kubelet) newVolumeUnmounterFromPlugins(kind string, name string, podUI
 }
 
+// newVolumeDetacherFromPlugins attempts to find a plugin by a name and then create a Detacher.
+// This function returns a valid volume Detacher or an error if the Detacher volume plugin was not found or if the Detacher failed to instantiate.
```
Same comments here as for `newVolumeAttacherFromPlugins`

pmorie commented on pkg/kubelet/volumes.go:
```diff
@@ -268,18 +257,17 @@ func (kl *Kubelet) newVolumeMounterFromPlugins(spec *volume.Spec, pod *api.Pod,
 	physicalMounter, err := plugin.NewMounter(spec, pod, opts)
 	if err != nil {
-		return nil, fmt.Errorf("failed to instantiate volume physicalMounter for %s: %v", spec.Name(), err)
+		return nil, fmt.Errorf("failed to instantiate physicalMounter for volume: %s using plugin: %s with a root cause: %v", spec.Name(), plugin.Name(), err)
```
I don't know what a 'physicalMounter' is -- just call it 'mounter' ?

pmorie commented on pkg/kubelet/volumes.go:
```diff
@@ -58,6 +56,9 @@ func (vh *volumeHost) GetKubeClient() clientset.Interface {
 
+// NewWrapperMounter attempts to create a volume mounter
+// from a volume Spec, pod and volume options.
+// Returns a new volume Mounter or an error
```
Missing period

pmorie commented on pkg/kubelet/volumes.go:
```diff
@@ -65,13 +66,12 @@ func (vh *volumeHost) NewWrapperMounter(volName string, spec volume.Spec, pod *a
 
+// NewWrapperUnmounter attempts to create a volume unmounter
+// from a volume name and pod uid.
+// Returns a new volume Unmounter or an error
```
Missing period, whole sentences here please.

pmorie commented on pkg/kubelet/volumes.go:
```diff
@@ -268,18 +254,21 @@ func (kl *Kubelet) newVolumeMounterFromPlugins(spec *volume.Spec, pod *api.Pod,
+//  - an attacher if one exists
+//  - an error if no plugin was found for the volume
+//    or the attacher failed to instantiate
+//  - nil if there is no appropriate attacher for this volume
```
This is kind of the exception to the whole sentence rule -- this is fine as is re: the specification of return values

pmorie commented on pkg/kubelet/volumes.go:
```diff
@@ -305,10 +297,7 @@ func (kl *Kubelet) newVolumeUnmounterFromPlugins(kind string, name string, podUI
 		// TODO: Maybe we should launch a cleanup of this dir?
 		return nil, fmt.Errorf("can't use volume plugins for %s/%s: %v", podUID, kind, err)
 	}
-	if plugin == nil {
```
This nil check has to go back in now, otherwise you will have a nil pointer when you call `plugin.NewUnmounter`

screeley44 replied:
@pmorie  - confused by this one...  FindPluginByName never returns a nil, nil - and the error check we do would catch any error scenario before we call the UnMounter - am I missing something here?

pmorie replied:
@screeley44 you're right, this check should be unnecessary now.

pmorie commented on pkg/kubelet/volumes.go:
```diff
@@ -268,18 +254,21 @@ func (kl *Kubelet) newVolumeMounterFromPlugins(spec *volume.Spec, pod *api.Pod,
 	if err != nil {
```
Also, please add godoc for `newVolumeMounterFromPlugins`

screeley44 commented:
The goal of this PR was to surface/expose plugin errors that were previously being eaten (i.e. errors that come from nfs.go, glusterfs.go, etc...).  And have them properly show in 'describe pod' so users could more easily see them without having to search the logs.  This accomplishes that, and all other errors that traditionally already showed in the 'describe' are not really affected.  After this I plan on continuing with additonal PRs and Issues to address other useability concerns in terms of error handling and other types of errors being eaten.

Below are a few example outputs from these changes, a newly surfaced plugin generated error that is now surfaced with true root cause and an existing non plugin type error that has always been surfaced and has not changed:

*Missing endpoints from GlusterFS using PV (previously being eaten):*
```
Events:
  FirstSeen	LastSeen	Count	From			SubobjectPath	Type		Reason		Message
  ---------	--------	-----	----			-------------	--------	------		-------
  15s		15s		1	{default-scheduler }			Normal		Scheduled	Successfully assigned bb-gluster-pod2 to 127.0.0.1
  15s		4s		2	{kubelet 127.0.0.1}			Warning		FailedMount	Unable to mount volumes for pod "bb-gluster-pod2_default(72a91ba0-fccb-11e5-9f13-52540092b5fb)": failed to instantiate mounter for volume: glustervol using plugin: kubernetes.io/persistent-claim with a root cause: failed to instantiate mounter for volume: pv-gluster using plugin: kubernetes.io/glusterfs with a root cause: endpoints "glusterfs-cluster" not found

kubelet log output:
E0407 10:17:23.715015   17134 glusterfs.go:83] glusterfs: failed to get endpoints glusterfs-cluster[endpoints "glusterfs-cluster" not found]
E0407 10:17:23.715030   17134 persistent_claim.go:85] Error creating mounter for claim: glusterfs-claim
E0407 10:17:23.715037   17134 volumes.go:125] Could not create volume mounter for pod 72a91ba0-fccb-11e5-9f13-52540092b5fb: failed to instantiate mounter for volume: glustervol using plugin: kubernetes.io/persistent-claim with a root cause: failed to instantiate mounter for volume: pv-gluster using plugin: kubernetes.io/glusterfs with a root cause: endpoints "glusterfs-cluster" not found
E0407 10:17:23.715056   17134 kubelet.go:1780] Unable to mount volumes for pod "bb-gluster-pod2_default(72a91ba0-fccb-11e5-9f13-52540092b5fb)": failed to instantiate mounter for volume: glustervol using plugin: kubernetes.io/persistent-claim with a root cause: failed to instantiate mounter for volume: pv-gluster using plugin: kubernetes.io/glusterfs with a root cause: endpoints "glusterfs-cluster" not found; skipping pod
E0407 10:17:23.715062   17134 pod_workers.go:138] Error syncing pod 72a91ba0-fccb-11e5-9f13-52540092b5fb, skipping: failed to instantiate mounter for volume: glustervol using plugin: kubernetes.io/persistent-claim with a root cause: failed to instantiate mounter for volume: pv-gluster using plugin: kubernetes.io/glusterfs with a root cause: endpoints "glusterfs-cluster" not found
I0407 10:17:23.715099   17134 server.go:608] Event(api.ObjectReference{Kind:"Pod", Namespace:"default", Name:"bb-gluster-pod2", UID:"72a91ba0-fccb-11e5-9f13-52540092b5fb", APIVersion:"v1", ResourceVersion:"149", FieldPath:""}): type: 'Warning' reason: 'FailedMount' Unable to mount volumes for pod "bb-gluster-pod2_default(72a91ba0-fccb-11e5-9f13-52540092b5fb)": failed to instantiate mounter for volume: glustervol using plugin: kubernetes.io/persistent-claim with a root cause: failed to instantiate mounter for volume: pv-gluster using plugin: kubernetes.io/glusterfs with a root cause: endpoints "glusterfs-cluster" not found
I0407 10:17:23.715140   17134 server.go:608] Event(api.ObjectReference{Kind:"Pod", Namespace:"default", Name:"bb-gluster-pod2", UID:"72a91ba0-fccb-11e5-9f13-52540092b5fb", APIVersion:"v1", ResourceVersion:"149", FieldPath:""}): type: 'Warning' reason: 'FailedSync' Error syncing pod, skipping: failed to instantiate mounter for volume: glustervol using plugin: kubernetes.io/persistent-claim with a root cause: failed to instantiate mounter for volume: pv-gluster using plugin: kubernetes.io/glusterfs with a root cause: endpoints "glusterfs-cluster" not found

```

*NFS Connection Timeout using PV (timeout error remains the same as this does not stem from plugins directly):*
```
Events:
  FirstSeen	LastSeen	Count	From			SubobjectPath	Type		Reason		Message
  ---------	--------	-----	----			-------------	--------	------		-------
  4m		4m		1	{default-scheduler }			Normal		Scheduled	Successfully assigned nfs-bb-pod2-pvc to 127.0.0.1
  2m		8s		2	{kubelet 127.0.0.1}			Warning		FailedMount	Unable to mount volumes for pod "nfs-bb-pod2-pvc_default(0304472c-fcca-11e5-9f13-52540092b5fb)": Mount failed: exit status 32
Mounting arguments: nfs1.rhs:/opt/data12 /var/lib/kubelet/pods/0304472c-fcca-11e5-9f13-52540092b5fb/volumes/kubernetes.io~nfs/pv-nfs nfs []
Output: mount.nfs: Connection timed out


kubelet log output:
I0407 10:07:06.898345   17134 kubelet.go:3242] Generating status for "nfs-bb-pod2-pvc_default(0304472c-fcca-11e5-9f13-52540092b5fb)"
I0407 10:07:06.911292   17134 manager.go:386] Status for pod "nfs-bb-pod2-pvc_default(0304472c-fcca-11e5-9f13-52540092b5fb)" updated successfully: {status:{Phase:Pending Conditions:[{Type:Ready Status:False LastProbeTime:{Time:{sec:0 nsec:0 loc:0x2e4da80}} LastTransitionTime:{Time:{sec:63595634826 nsec:0 loc:0x2e70c60}} Reason:ContainersNotReady Message:containers with unready status: [nfs-bb-pod2-pvc]}] Message: Reason: HostIP:127.0.0.1 PodIP: StartTime:0xc2089fde60 ContainerStatuses:[{Name:nfs-bb-pod2-pvc State:{Waiting:0xc2089fde00 Running:<nil> Terminated:<nil>} LastTerminationState:{Waiting:<nil> Running:<nil> Terminated:<nil>} Ready:false RestartCount:0 Image:busybox ImageID: ContainerID:}]} version:1 podName:nfs-bb-pod2-pvc podNamespace:default}
E0407 10:09:17.994381   17134 kubelet.go:1780] Unable to mount volumes for pod "nfs-bb-pod2-pvc_default(0304472c-fcca-11e5-9f13-52540092b5fb)": Mount failed: exit status 32
Mounting arguments: nfs1.rhs:/opt/data12 /var/lib/kubelet/pods/0304472c-fcca-11e5-9f13-52540092b5fb/volumes/kubernetes.io~nfs/pv-nfs nfs []
Output: mount.nfs: Connection timed out

; skipping pod
E0407 10:09:17.994399   17134 pod_workers.go:138] Error syncing pod 0304472c-fcca-11e5-9f13-52540092b5fb, skipping: Mount failed: exit status 32
Mounting arguments: nfs1.rhs:/opt/data12 /var/lib/kubelet/pods/0304472c-fcca-11e5-9f13-52540092b5fb/volumes/kubernetes.io~nfs/pv-nfs nfs []
Output: mount.nfs: Connection timed out

I0407 10:09:17.994429   17134 server.go:608] Event(api.ObjectReference{Kind:"Pod", Namespace:"default", Name:"nfs-bb-pod2-pvc", UID:"0304472c-fcca-11e5-9f13-52540092b5fb", APIVersion:"v1", ResourceVersion:"68", FieldPath:""}): type: 'Warning' reason: 'FailedMount' Unable to mount volumes for pod "nfs-bb-pod2-pvc_default(0304472c-fcca-11e5-9f13-52540092b5fb)": Mount failed: exit status 32
Mounting arguments: nfs1.rhs:/opt/data12 /var/lib/kubelet/pods/0304472c-fcca-11e5-9f13-52540092b5fb/volumes/kubernetes.io~nfs/pv-nfs nfs []
Output: mount.nfs: Connection timed out

I0407 10:09:17.994468   17134 server.go:608] Event(api.ObjectReference{Kind:"Pod", Namespace:"default", Name:"nfs-bb-pod2-pvc", UID:"0304472c-fcca-11e5-9f13-52540092b5fb", APIVersion:"v1", ResourceVersion:"68", FieldPath:""}): type: 'Warning' reason: 'FailedSync' Error syncing pod, skipping: Mount failed: exit status 32
Mounting arguments: nfs1.rhs:/opt/data12 /var/lib/kubelet/pods/0304472c-fcca-11e5-9f13-52540092b5fb/volumes/kubernetes.io~nfs/pv-nfs nfs []
Output: mount.nfs: Connection timed out

I0407 10:09:29.817521   17134 kubelet.go:3242] Generating status for "nfs-bb-pod2-pvc_default(0304472c-fcca-11e5-9f13-52540092b5fb)"
I0407 10:09:29.817652   17134 manager.go:277] Ignoring same status for pod "nfs-bb-pod2-pvc_default(0304472c-fcca-11e5-9f13-52540092b5fb)", status: {Phase:Pending Conditions:[{Type:Ready Status:False LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2016-04-07 10:07:06 -0400 EDT Reason:ContainersNotReady Message:containers with unready status: [nfs-bb-pod2-pvc]}] Message: Reason: HostIP:127.0.0.1 PodIP: StartTime:2016-04-07 10:07:06 -0400 EDT ContainerStatuses:[{Name:nfs-bb-pod2-pvc State:{Waiting:0xc208ab6740 Running:<nil> Terminated:<nil>} LastTerminationState:{Waiting:<nil> Running:<nil> Terminated:<nil>} Ready:false RestartCount:0 Image:busybox ImageID: ContainerID:}]}
E0407 10:11:40.906401   17134 kubelet.go:1780] Unable to mount volumes for pod "nfs-bb-pod2-pvc_default(0304472c-fcca-11e5-9f13-52540092b5fb)": Mount failed: exit status 32
Mounting arguments: nfs1.rhs:/opt/data12 /var/lib/kubelet/pods/0304472c-fcca-11e5-9f13-52540092b5fb/volumes/kubernetes.io~nfs/pv-nfs nfs []
Output: mount.nfs: Connection timed out

```


@pmorie

pmorie commented:
@screeley44 I would change the title of this PR (since it will become an entry in the release notes file) to be something like 'Return more useful error information when a persistent volume fails to mount'.

@saad-ali Do you have bandwidth to give this a review and then one of us can tag?

saad-ali commented:
> @saad-ali Do you have bandwidth to give this a review and then one of us can tag?

Will do.

saad-ali commented:
LGTM

pmorie commented:
Thanks for the patch @screeley44.  I think this is one of the big usability issues w/ PVs today.

markturansky commented:
>
> Thanks for the patch @screeley44 <https://github.com/screeley44>. I think
> this is one of the big usability issues w/ PVs today.
>

+1

screeley44 commented:
@pmorie  -  am I supposed to do something with the failed Jenkins?  I have not found the cause of the failure yet

screeley44 commented:
@k8s-bot test this issue: #24148

swagiaal commented:
@k8s-bot test this issue: #24148
